contributors/design-proposals/auth/kubectl-exec-plugins.md

```diff
@@ -0,0 +1,193 @@
+# Out-of-tree client authentication providers
+
+Author: @ericchiang
+
+# Objective
+
+This document describes a credential rotation strategy for kubectl and client-go
+using an exec-based plugin mechanism.
+
+# Background
+
+Kubernetes clients can provide three kind of credentials: bearer tokens, TLS
+client certs, and basic auth username and password. Kubeconfigs can either
+in-line the credential, load credentials from a file, or can use an `AuthProvider`
+to actively fetch and rotate credentials. `AuthProviders` are compiled into kubectl
+and target specific providers (GCP, Keystone, Azure AD) or implement a
+specification supported but a subset of vendors (OpenID Connect).
+
+External tools that do rotation generally have to be called on a regular bases in
+case the current credentials have expired, which leads to wrapping `kubectl` with
+another tool or aliasing. For example, this is a [real example](
+https://github.com/heptio/authenticator#4-set-up-kubectl-to-use-heptio-authenticator-for-aws-tokens)
+from Heptio's AWS authenticator:
+
+```
+kubectl --kubeconfig /path/to/kubeconfig --token "$(heptio-authenticator-aws token -i CLUSTER_ID -r ROLE_ARN)" [...]
+```
+
+Beside resulting in a long command, this potentially encourages distributions to
+wrap or fork kubectl, changing the way that users interact with different
+Kubernetes clusters.
+
+# Proposal
+
+This proposal builds off of earlier requests to [support exec-based plugins](
+https://github.com/kubernetes/kubernetes/issues/35530#issuecomment-256170024), and
+proposes that we should add this as a first-class feature of kubectl. Specifically,
+kubectl (and client-go) should be able to receive credentials by executing a
+command and reading that command's stdout.
+
+In fact, kubectl already does this today. The GCP plugin can already be configured
+to [call a command](
+https://github.com/kubernetes/client-go/blob/kubernetes-1.8.5/plugin/pkg/client/auth/gcp/gcp.go#L228-L240)
+other than `gcloud`.
+
+## Kubeconfig changes
+
+The current `AuthProviderConfig` uses `map[string]string` for configuration, which
+makes it hard to express things like a list of arguments or list key/value environment
+variables. As such, `AuthProviderConfig` should add another field which expresses the
+`exec` config. This has the benefit of a more natural structure, but the trade-off of
+not being compatible with the existing `kubectl config set-credentials` implementation.
+
+```go
+const (
+    AuthProviderExecKindBearerToken = "bearer-token"
+    AuthProviderExecKindClientCert  = "tls-certificate"
+)
+
+type AuthProviderConfig struct {
+    Name   string            `json:"name"`
+    Config map[string]string `json:"config"`
+
+    // Exec should only be provided if Name and Config aren't.
```

Comment:
If `Exec` and (`Name`,`Config`) are mutually exclusive, would it make more sense to move the `Exec` config up a level to `AuthInfo` instead? It makes some sense to me to add a "dynamic" token (or cert) option along side `Token` and `TokenFile`.

I don't mean to over complicate the issue, it just seems a bit complicated from a user's perspective. It would make more sense to me if either this were its own named auth provider (e.g. `Name: "exec"`), moved out of the auth provider section, or allowed to specify a named auth provider as well with some sort of defined login on how that case is handled.

Reply:
Agreed this is an awkward place. I'll move it a level up.

Reply:
Addressed